File: 백준/Gold/15686. 치킨 배달/치킨 배달.cc
#include <bits/stdc++.h>
using namespace std;

int n, m, ret = 1e9, a[54][54];
vector<pair<int, int>> chicken, house;

int chicken_dist(vector<int> &v) {
  int ret = 0;
  for (int i = 0; i < house.size(); i++) {
    int mn = 1e9;
    for (int j = 0; j < v.size(); j++) {
      int dist = abs(house[i].first - chicken[v[j]].first) +
                 abs(house[i].second - chicken[v[j]].second);
      mn = min(mn, dist);
    }
    ret += mn;
  }

  return ret;
}

void go(int start, vector<int> &v) {
  if (v.size() == m) {
    ret = min(ret, chicken_dist(v));
    return;
  }

  for (int i = start + 1; i < chicken.size(); i++) {
    v.push_back(i);
    go(i, v);
    v.pop_back();
  }

  return;
}

int main() {
  cin >> n >> m;

  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      cin >> a[i][j];
      if (a[i][j] == 1)
        house.push_back({i, j});
      else if (a[i][j] == 2)
        chicken.push_back({i, j});
    }
  }

  vector<int> v;

  go(-1, v);

  cout << ret;

  return 0;
}

// 치킨 거리가 가장 작아지는 M개 치킨집 고르기
// 재귀를 통한 완탐
//

File: 백준/Gold/15686. 치킨 배달/치킨 배달.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : './input.txt').toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const a = input.slice(1).map(line => line.split(' ').map(Number));

const _house = [], _chicken = [];

let ret = 1e9;

a.forEach((line, i) => line.forEach((e, j) => {
  if(e === 1){
    _house.push([i, j]);
  }else if(e ===2){
    _chicken.push([i, j]);
  }
}))

const chicken_dist = (v) => {
  let ret = 0;
  for (let i = 0; i < _house.length; i++){
    let mn = 1e9;
    for (let j = 0; j < v.length; j++){
      let dist = Math.abs(_house[i][0] - _chicken[v[j]][0]) + Math.abs(_house[i][1] - _chicken[v[j]][1]);
      mn = Math.min(dist, mn);
    }
    ret += mn;
  }
  return ret;
}

const go = (a, v = []) => {
  if(v.length === m){
    // logic
    ret = Math.min(ret, chicken_dist(v));
  }

  for (let i = a + 1; i < _chicken.length; i++){
    v.push(i);
    go(i, v);
    v.pop();
  }
}

const answer = () => {
  go(-1);

  console.log(ret);
  
}

answer();
